dxr/build.py

```diff
@@ -370,9 +367,19 @@ def file_contents(path, encoding_guess):  # TODO: Make accessible to TreeToIndex
     """
     # Read the binary contents of the file.
     with open(path, 'rb') as source_file:
-        contents = source_file.read()  # always str
-    _, contents = decode_data(contents, encoding_guess)
-    return contents  # unicode if we were able to decode, str if not
+        try:
+            # Create a read-only mmap of the entire file.
+            mmap_block = mmap.mmap(source_file.fileno(), 0, prot=mmap.PROT_READ, flags=mmap.MAP_PRIVATE)
+            decoded, contents = decode_block(mmap_block, encoding_guess)
+            if not decoded and force_read:
+                mmap_block.seek(0)
+                contents = str(mmap_block)
+            mmap_block.close()
+            # unicode if we were able to decode, str if not
+            return contents
+        except ValueError:
```

Comment:
Does the exception occur on the initial mmap() call? Can we tighten up this exception block at all? I worry because ValueError is pretty generic.